clang-tools-extra/clang-tidy/bugprone/LoopVariableCopiedThenModifiedCheck.cpp

```diff
@@ -0,0 +1,95 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "LoopVariableCopiedThenModifiedCheck.h"
+#include "../utils/Matchers.h"
+#include "../utils/TypeTraits.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Analysis/Analyses/ExprMutationAnalyzer.h"
+#include "clang/Basic/Diagnostic.h"
+#include "clang/Lex/Lexer.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+namespace {
+AST_MATCHER(VarDecl, isInMacro) { return Node.getBeginLoc().isMacroID(); }
+} // namespace
+
+LoopVariableCopiedThenModifiedCheck::LoopVariableCopiedThenModifiedCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context), IgnoreInexpensiveVariables(Options.get(
+                                         "IgnoreInexpensiveVariables", false)),
+      WarnOnlyOnAutoCopies(Options.get("WarnOnlyOnAutoCopies", false)) {}
+
+void LoopVariableCopiedThenModifiedCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "IgnoreInexpensiveVariables", IgnoreInexpensiveVariables);
+  Options.store(Opts, "WarnOnlyOnAutoCopies", WarnOnlyOnAutoCopies);
+}
+
+void LoopVariableCopiedThenModifiedCheck::registerMatchers(
+    MatchFinder *Finder) {
+  const auto HasReferenceOrPointerType = hasType(qualType(
+      unless(hasCanonicalType(anyOf(referenceType(), pointerType())))));
+  const auto IteratorReturnsValueType = cxxOperatorCallExpr(
+      hasOverloadedOperatorName("*"),
+      callee(
+          cxxMethodDecl(returns(unless(hasCanonicalType(referenceType()))))));
+  const auto NotConstructedByCopy = cxxConstructExpr(
+      hasDeclaration(cxxConstructorDecl(unless(isCopyConstructor()))));
+  const auto ConstructedByConversion =
+      cxxMemberCallExpr(callee(cxxConversionDecl()));
+  const auto LoopVar =
+      varDecl(unless(isInMacro()), HasReferenceOrPointerType,
+              unless(hasInitializer(expr(hasDescendant(expr(
+                  anyOf(materializeTemporaryExpr(), IteratorReturnsValueType,
+                        NotConstructedByCopy, ConstructedByConversion)))))));
+  Finder->addMatcher(cxxForRangeStmt(hasLoopVariable(LoopVar.bind("loopVar")))
```

Comment:
consider excluding system headers or implicit code like template instances, otherwise there could be conflicts on fixes